### DISCUSSION TITLE
How to know if isNullable from VariableElement?

### DISCUSSION THREAD

**Discussion by universezy** [square/kotlinpoet#1596] [Q&A] [answered]
I got ExecutableElement from method, then I got list of VariableElement from ExecutableElement.parameters, for each VariableElement, how can I get that if the param is nullable with symbol "?" ?
If I try to "asType" or "asTypeName" for VariableElement, it returns a typeName with "isNullable" false always.
Please help me

**Comment by ZacSweers** [comment, marked as answer]
You would have to read the kotlin metadata from the enclosing class, find that corresponding KmFunction, and finally the KmParameter on that function.

**Reply by universezy** [reply]
Thx too much, I'll try it.

**Reply by universezy** [reply]
Could you pls tell me how to read the kotlin metadata from the VariableElement? I try to cast the VariableElement to the Symbol.MethodSymbol, but it throws exception"Symbol is declared in module 'jdk.compiler' which does not export package 'com.sun.tools.javac.code'", as you can see, I can't visit code in javac, btw it's a lib module in my android project.

**Reply by ZacSweers** [reply]
That would be a good exercise to the reader. We have a lot of docs on the Kotlinpoet docs site for reading metadata.

**Reply by universezy** [reply]
I realized may be I had expressed the wrong meaning to you, I try to get informations from an existed method and create code with them by kotlinpoet, but now I realize that it's just a JAVA question: Where is the "nullable" info while compiling kotlin files, so I see the compiled class file, and find that the "nullable" symbol was converted to the "Nullable" annotation, so I can use "Element.getAnnotation(Nullable::class.java) != null" to get "isNullable" value.

**Reply by ZacSweers** [reply]
That's not a good way to do it, you should use metadata to be absolutely correct. You have no guarantee that that annotation will be there at runtime, nor does it propagate correctly with type arguments on generic functions. I'm unmarking that as an answer because we don't recommend users do that.

**Reply by universezy** [reply]
I got an ExecutableElement from the method in an existed Interface file and a list of VariableElement from the ExecutableElement, but there is no api to get any metadata, how can I do? Could you pls give some better suggestions? I had searched for it in docs and read source codes, but no idea.

**Reply by universezy** [reply]
> You would have to read the kotlin metadata from the enclosing class, find that corresponding KmFunction, and finally the KmParameter on that function.

There is a property named "metadata" in class "com.sun.tools.javac.code.Symbol", did you mean it?
I guess I misunderstood what you said.

**Reply by ZacSweers** [reply]
I think you should look more closely, as it's the first result for "metadata" on the docs site

https://square.github.io/kotlinpoet/interop-kotlinx-metadata/

<img width="983" alt="image" src="https://github.com/square/kotlinpoet/assets/1361086/f8fd83f4-8dc5-4b5e-8ffa-ff126f53bb02">

**Reply by universezy** [reply]
Ok, thx, I misunderstood it. I'll read it and bring the lib into my project.
